File: super-tiny-interpreter-v1/src/scripts.ts
import { ASTNode, ASTNodeType } from "./models";

/**
 * 遍历AST,计算
 */

export default class Script {
  static verbose: boolean = false;
  variables: Map<string, number> = new Map();

  evaluate(node: ASTNode, indent: string): number | null {
    let result: number = null;
    switch (node.type) {
      case ASTNodeType.Program:
        for (let child of node.children) {
          result = this.evaluate(child, indent)
        }
        break;
      case ASTNodeType.Additive: {
        let child1 = node.children[0];
        let val1 = this.evaluate(child1, `${indent}\t`);
        let child2 = node.children[1];
        let val2 = this.evaluate(child2, `${indent}\t`);
        if (node.text == "+") result = val1 + val2;
        else result = val1 - val2;
        break;
      }
      case ASTNodeType.Multiplicative: {
        let child1 = node.children[0];
        let val1 = this.evaluate(child1, `${indent}\t`);
        let child2 = node.children[1];
        let val2 = this.evaluate(child2, `${indent}\t`);
        if (node.text == "*") result = val1 * val2;
        else result = val1 / val2;
        break;
      }
      case ASTNodeType.IntLiteral:
        result = parseInt(node.text);
        break;
      case ASTNodeType.Identifier: {
        let varName = node.text;
        if (this.variables.has(varName)) {
          let value = this.variables.get(varName);
          if (!value) throw Error(`variable ${varName} has not been set any value`)
          result = value;
        } else throw Error(`unknown variable: ${varName}`)
        break;
      }
      case ASTNodeType.AssignmentStmt: {
        let varName = node.text;
        if (!this.variables.has(varName)) throw Error(`unknown variable: ${varName}`)
        //继续给变量赋值
      }
      case ASTNodeType.IntDeclaration: {
        let varName = node.text;
        let varValue = null;
        if (node.children.length) { //说明是初始化语句，给变量赋值
          let child = node.children[0];
          varValue = this.evaluate(child, `${indent}\t`);
        }
        this.variables.set(varName, varValue);
        result = varValue;
        break;
      }
    }
    if (Script.verbose) {
      console.log(`${indent} Result: ${result}`)
    }
    return result;
  }
}

File: cpp-super-tiny-interpreter/src/scripts.cpp
#include "scripts.h"
int Script::evaluate(ASTNode &node, string indent)
{
  int result;
  switch (node.type)
  {
  case ASTNodeType::Program:
    for (list<ASTNode>::iterator iter = node.children.begin(); iter != node.children.end(); ++iter)
    {
      result = evaluate(*iter, indent);
    }
    break;
  case ASTNodeType::Additive:
  {
    auto iter = node.children.begin();
    auto child1 = *iter;
    int val1 = evaluate(child1, indent + "\t");
    ++iter;
    auto child2 = *iter;
    int val2 = evaluate(child2, indent + "\t");
    if (node.text == "+")
      result = val1 + val2;
    else
      result = val1 - val2;
    break;
  }
  case ASTNodeType::Multiplicative:
  {
    auto iter = node.children.begin();
    auto child1 = *iter;
    int val1 = evaluate(child1, indent + "\t");
    ++iter;
    auto child2 = *iter;
    int val2 = evaluate(child2, indent + "\t");
    if (node.text == "*")
      result = val1 * val2;
    else
      result = val1 / val2;
    break;
  }
  case ASTNodeType::IntLiteral:
    result = stoi(node.text);
    break;
  case ASTNodeType::Identifier:
  {
    string varName = node.text;
    try
    {
      int value = variables.at(varName);
      result = value;
    }
    catch (exception e)
    {
      throw("unknown variable" + varName);
    }
    break;
  }
  case ASTNodeType::AssignmentStmt:
  {
    auto varName = node.text;
    try
    {
      variables.at(varName);
    }
    catch (exception e)
    {
      throw "unknown variable" + varName;
    }
    //继续给变量赋值
  }
  case ASTNodeType::IntDeclaration:
  {
    auto varName = node.text;
    int varValue;
    if (node.children.size())
    { //说明是初始化语句，给变量赋值
      auto child = node.children.front();
      varValue = evaluate(child, indent + "\t");
    }
    variables.erase(varName);
    variables.emplace(varName, varValue);
    result = varValue;
    break;
  }
  }
  if (verbose)
  {
    cout << indent + "Result:" << result << endl;
  }
  return result;
}
